File: apps/bff/src/main/java/com/example/demo/model/UserDocument.java
package com.example.demo.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.io.Serializable;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Document metadata stored in MongoDB
 * Actual files stored in S3
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "documents")
public class UserDocument implements Serializable {

    @Id
    private String documentId;

    // File information
    private String fileName; // Sanitized filename
    private String originalFileName; // Original user-provided filename
    private Long fileSize; // Bytes
    private String contentType; // MIME type
    private DocumentCategory category;

    // S3 location
    private String s3Key; // Full S3 path
    private String s3Bucket;

    // Ownership - whose document this is
    @Indexed
    private String ownerIdType; // HSID, EID, OHID, MSID
    @Indexed
    private String ownerIdValue; // The actual ID value
    private String ownerFirstName;
    private String ownerLastName;

    // Upload information - who uploaded it
    @Indexed
    private String uploadedByIdType;
    @Indexed
    private String uploadedByIdValue;
    private String uploadedByFirstName;
    private String uploadedByLastName;
    @Indexed
    private Instant uploadedAt;

    // Security and access
    @Builder.Default
    private Boolean isSensitive = true; // Default: all documents are sensitive

    // Metadata for search
    private String description;
    @Builder.Default
    private List<String> tags = new ArrayList<>();
    private String extractedText; // Future: OCR text for search

    // Status
    @Indexed
    private DocumentStatus status;
    @Indexed
    private Instant finalizedAt; // When user clicked "Upload" button
    private Instant deletedAt;

    // Temporary upload tracking
    @Indexed
    private String sessionId; // For temp files cleanup
    private String tempS3Key; // Temporary S3 location before finalization

    // Antivirus scan status
    @Indexed
    private AntivirusStatus avStatus;
    private String virusName; // If infected, name of virus
    private Instant avScannedAt; // When AV scan completed

    // Audit
    private Instant lastModifiedAt;
    private Instant lastAccessedAt;

    public enum AntivirusStatus {
        PENDING,      // Waiting for AV scan
        SCANNING,     // Currently being scanned
        CLEAN,        // Passed AV scan
        INFECTED,     // Failed AV scan - contains malware
        SCAN_ERROR    // AV scan failed due to error
    }

    public enum DocumentCategory {
        MEDICAL_RECORD,
        LAB_RESULT,
        PRESCRIPTION,
        INSURANCE_CARD,
        IMMUNIZATION_RECORD,
        MARKSHEET,
        IDENTIFICATION,
        OTHER
    }

    public enum DocumentStatus {
        TEMPORARY,    // Uploaded but not finalized
        ACTIVE,       // Finalized and visible
        DELETED       // Soft deleted
    }

    /**
     * Get display name for the owner
     */
    public String getOwnerDisplayName() {
        if (ownerFirstName != null && ownerLastName != null) {
            return ownerFirstName + " " + ownerLastName;
        }
        return ownerIdValue;
    }

    /**
     * Get display name for uploader
     */
    public String getUploadedByDisplayName() {
        if (uploadedByFirstName != null && uploadedByLastName != null) {
            return uploadedByFirstName + " " + uploadedByLastName;
        }
        return uploadedByIdValue;
    }

    /**
     * Check if document is temporary (not finalized)
     */
    public boolean isTemporary() {
        return DocumentStatus.TEMPORARY.equals(status);
    }

    /**
     * Check if document is active and viewable
     */
    public boolean isActive() {
        return DocumentStatus.ACTIVE.equals(status) && deletedAt == null;
    }
}

File: apps/mfe-documents/src/types/userDocument.ts
/**
 * Document types matching backend models
 */

export enum DocumentCategory {
  MEDICAL_RECORD = 'MEDICAL_RECORD',
  LAB_RESULT = 'LAB_RESULT',
  PRESCRIPTION = 'PRESCRIPTION',
  INSURANCE_CARD = 'INSURANCE_CARD',
  IMMUNIZATION_RECORD = 'IMMUNIZATION_RECORD',
  MARKSHEET = 'MARKSHEET',
  IDENTIFICATION = 'IDENTIFICATION',
  OTHER = 'OTHER',
}

export enum DocumentStatus {
  TEMPORARY = 'TEMPORARY',
  ACTIVE = 'ACTIVE',
  DELETED = 'DELETED',
}

export interface Document {
  documentId: string;
  fileName: string;
  originalFileName: string;
  fileSize: number;
  contentType: string;
  category: DocumentCategory;
  s3Key: string;
  s3Bucket: string;

  // Ownership
  ownerIdType: string;
  ownerIdValue: string;
  ownerFirstName?: string;
  ownerLastName?: string;

  // Upload info
  uploadedByIdType: string;
  uploadedByIdValue: string;
  uploadedByFirstName?: string;
  uploadedByLastName?: string;
  uploadedAt: string;

  // Security
  isSensitive: boolean;

  // Metadata
  description?: string;
  tags?: string[];
  extractedText?: string;

  // Status
  status: DocumentStatus;
  finalizedAt?: string;
  deletedAt?: string;

  // Audit
  lastModifiedAt?: string;
  lastAccessedAt?: string;
}

export interface FileUploadInfo {
  fileName: string;
  contentType: string;
  fileSize: number;
}

export interface DocumentUploadRequest {
  files: FileUploadInfo[];
  ownerIdType: string;
  ownerIdValue: string;
}

export interface UploadInfo {
  tempDocumentId: string;
  fileName: string;
  presignedUrl: string;
  s3Key: string;
  expiresIn: number;
}

export interface DocumentUploadResponse {
  uploads: UploadInfo[];
}

export interface DocumentFinalizeRequest {
  tempDocumentIds: string[];
  category: DocumentCategory;
  description?: string;
  tags?: string[];
}

export interface DocumentSearchRequest {
  ownerIdType: string;
  ownerIdValue: string;
  searchQuery?: string;
  categories?: DocumentCategory[];
  uploadedAfter?: string;
  uploadedBefore?: string;
  tags?: string[];
  includeSensitive?: boolean;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

export interface DocumentSearchResponse {
  content: Document[];
  pageable: {
    pageNumber: number;
    pageSize: number;
  };
  totalElements: number;
  totalPages: number;
  last: boolean;
  first: boolean;
}

export interface DocumentDownloadResponse {
  downloadUrl: string;
  expiresIn: number;
}

export interface CategoryOption {
  value: string;
  label: string;
}

export const CATEGORY_LABELS: Record<DocumentCategory, string> = {
  [DocumentCategory.MEDICAL_RECORD]: 'Medical Record',
  [DocumentCategory.LAB_RESULT]: 'Lab Result',
  [DocumentCategory.PRESCRIPTION]: 'Prescription',
  [DocumentCategory.INSURANCE_CARD]: 'Insurance Card',
  [DocumentCategory.IMMUNIZATION_RECORD]: 'Immunization Record',
  [DocumentCategory.MARKSHEET]: 'Marksheet / Transcript',
  [DocumentCategory.IDENTIFICATION]: 'Identification',
  [DocumentCategory.OTHER]: 'Other',
};

export const MAX_FILE_SIZE = 25 * 1024 * 1024; // 25 MB
export const MAX_FILES_PER_UPLOAD = 5;
export const ALLOWED_FILE_TYPES = [
  'application/pdf',
  'image/jpeg',
  'image/jpg',
  'image/png',
  'image/gif',
  'image/webp',
  'application/msword',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
];
